src/models/promise_queue.ts

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import EventEmitter from 'events';
+import PriorityQueue from './priority_queue';
+
+type TaskOptions = { signal?: AbortSignal };
+
+type Task<TaskResultType> =
+  | ((options?: TaskOptions) => PromiseLike<TaskResultType>)
+  | ((options?: TaskOptions) => TaskResultType);
+
+type AddTaskOptions = {
+  priority?: number;
+  signal?: AbortSignal;
+};
+
+export default class PromiseQueue extends EventEmitter {
+  #queue: PriorityQueue<() => Promise<void>>;
+
+  // Number of active jobs currently running.
+  #jobsRunning: number;
+
+  #allowedConcurrentJobs: number = 1;
+
+  #isPaused = false;
+
+  #intervalId: ReturnType<typeof setTimeout> | null = null;
+
+  constructor() {
+    super();
+    this.#queue = new PriorityQueue<() => Promise<void>>();
+    this.#jobsRunning = 0;
+    this.#startInterval();
+  }
+
+  #startInterval() {
+    this.processQueue();
+    this.#intervalId = setTimeout(() => {
+      if (this.#isPaused) {
+        this.#clearInterval();
+        return;
+      }
+      this.#startInterval();
+    }, 1000);
```

Comment:
The interval has 2 reasons for existing, if somehow the tasks finishes but cannot start another the next time the interval runs it will start tasks until we reach max concurrency.

There is another reason which has not been implemented yet which is for rate limiting tasks.
We could set the interval and only allow X tasks to run in until the interval is over, since we do not require this for the GLL I did not implement it but I will implement this on another PR.

Since the first reason does not require configuring the interval time I made an informed guess so tasks don't wait more than 1 second to run if its possible to run them.

Reply:
This is useful to implement the bucket token algorithm for rate-limit requests, indeed.